src/utils.h

```diff
@@ -0,0 +1,148 @@
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+#ifndef SGDNET_UTILS_
+#define SGDNET_UTILS_
+
+#include <RcppArmadillo.h>
+#include "families.h"
+
+//' Computes the squared norm over rows
+//'
+//' @param x arma::mat or arma::sp_mat
+//'
+//' @return The maximum rowise sum of x.
+template <typename T>
+double ColNormsMax(const T& x) {
+  return arma::sum(arma::square(x), 0).max();
+}
+
+//' Retrieve step size for the chosen family and feature matrix X
+//'
+//' @param max_squared_sum the maximum squared rowise norm from X
+//' @param alpha_scaled alpha/n_samples
+//' @param family the model family
+//' @param fit_intercept should an intercept be added to the model?
+//' @param n_samples the number of samples in X
+//'
+//' @return Step size.
+//'
+//' @keywords internal
+double GetStepSize(double         max_squared_sum,
+                   double         alpha_scaled,
+                   sgdnet::Family family,
+                   bool           fit_intercept,
+                   arma::uword    n_samples) {
+  double L;
+
+  switch(family) {
+    case sgdnet::GAUSSIAN: L = max_squared_sum + fit_intercept + alpha_scaled;
+                           break;
+  }
+
+  return 1.0 / (2.0*L + std::min(2.0*n_samples*alpha_scaled, L));
+}
+
+//' Return index of nonzero elements
+//'
+//' The function is overloaded to work with both dense and sparse matrices.
+//'
+//' @param x_it iterator
+//' @param i counter
+//'
+//' @return Index of iterator in current row or column.
+//'
+//' @keywords internal
+arma::uvec Nonzeros(const arma::sp_mat& x, arma::uword i) {
+
+  arma::sp_mat::const_col_iterator x_itr = x.begin_col(i);
+  arma::sp_mat::const_col_iterator x_end = x.end_col(i);
+
+  std::vector<unsigned int> out;
+
+  for(; x_itr != x_end; ++x_itr) {
+    out.push_back(x_itr.row());
+  }
+
+  return arma::conv_to< arma::uvec >::from(out);
+}
+
+// For dense matrices, just return a sequence of integers
+// TODO: consider using a conditional for sparse matrices to avoid
+//       entering this function.
+arma::uvec Nonzeros(const arma::mat& x, arma::uword i) {
+  return arma::regspace<arma::uvec>(0, x.n_rows - 1);
+}
+
+//' Preprocess data
+//'
+//' @param x feature matrix, sparse or dense
+//' @param y targets
+//' @param normalize whether to normalize x
+//' @param fit_intercept wheter to fit the intercept. If false, no
+//'   scaling or centering is done.
+//' @param x_offset a vector of offsets for each feature
+//' @param x_scale a vector of scaling factors (l2 norms) for each vector
+//' @param y_offset a vector of means for each column in y
+//'
+//' @return Nothing. x and y are scaled and centered.
+//' @keywords internal
+template <typename T>
+void Preprocess(T&            x,
+                arma::mat&    y,
+                const bool    normalize,
+                const bool    fit_intercept,
+                arma::vec&    x_offset,
```

Comment:
Can we use a different term than "offset"? 

"Offset" has a statistical meaning (fixed term added to linear predictor in GLM) and IIRC we plan on supporting that sense of "offset" eventually.

Reply:
Either works. When we add observation weights eventually, this might need to be a weighted mean (not sure what `glmnet` does), so "center" might be less ambiguous.